src/main/java/org/folio/fqm/service/EntityTypeFlatteningService.java

```diff
@@ -0,0 +1,242 @@
+package org.folio.fqm.service;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import kotlin.Pair;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+import org.folio.fqm.exception.EntityTypeNotFoundException;
+import org.folio.fqm.repository.EntityTypeRepository;
+import org.folio.querytool.domain.dto.EntityType;
+import org.folio.querytool.domain.dto.EntityTypeColumn;
+import org.folio.querytool.domain.dto.EntityTypeSource;
+import org.folio.querytool.domain.dto.EntityTypeSourceJoin;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+
+@Service
+@RequiredArgsConstructor
+@Log4j2
+public class EntityTypeFlatteningService {
+  private final EntityTypeRepository entityTypeRepository;
+  private final ObjectMapper objectMapper;
+
+  // TODO: clean up
+  public Optional<EntityType> getFlattenedEntityType(UUID entityTypeId, boolean doFinalRenames) {
+    EntityType originalEntityType = entityTypeRepository
+      .getEntityTypeDefinition(entityTypeId)
+      .orElseThrow(() -> new EntityTypeNotFoundException(entityTypeId));
+    EntityType flattenedEntityType = new EntityType()
+      .id(originalEntityType.getId())
+      .name(originalEntityType.getName())
+      ._private(originalEntityType.getPrivate())
+      .defaultSort(originalEntityType.getDefaultSort())
+      .columns(originalEntityType.getColumns()) // TODO: probably remove
+      .idView(originalEntityType.getIdView())
+      .customFieldEntityTypeId(originalEntityType.getCustomFieldEntityTypeId())
+      .labelAlias(originalEntityType.getLabelAlias())
+      .root(originalEntityType.getRoot())
+      .sourceView(originalEntityType.getSourceView()) // Possibly unneeded
+      .sourceViewExtractor(originalEntityType.getSourceViewExtractor()); // Possibly unneeded
+
+    List<EntityTypeColumn> finalColumns = new ArrayList<>();
+    for (EntityTypeSource source : originalEntityType.getSources()) {
+      if (source.getType().equals("db")) {
+        Pair<EntityTypeSource, List<EntityTypeColumn>> updatePair = getConvertedSourceAndColumns(originalEntityType, source, null, false); // TODO: think about this, may not be able to hardcode false here
```

Comment:
Comments like this make me wish we _were_ using something where you can name parameters at the call-site, like kotlin. I have no idea what the false does here, which is something that could be fixed at the language-level if Java supported it ☹️